core/src/main/java/nu/yona/server/goals/entities/IGoal.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2020 Stichting Yona Foundation This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v.2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+package nu.yona.server.goals.entities;
+
+import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
+import java.time.temporal.TemporalUnit;
+import java.util.Optional;
+
+import nu.yona.server.analysis.entities.DayActivity;
+import nu.yona.server.util.TimeUtil;
+
+public interface IGoal
```

Comment:
Does the introduction of those interfaces not complicate the architecture? So now we have entities, DTOs and also interfaces that make both quite overlapping. Isn't it cleaner to have the logic in one place instead, and not have interfaces that span across both entities and DTOs (which are two layers in the architecture)?

Reply:
This was introduced through this commit: 45a7f320ce78f4e11fd748b1b7e621b131ab8db5. Your comment precisely touches the issue. The optimization was to use the cached goal rather than fetching the goal entities from the database again. Given we don't want to use the DTOs in the entity layer, I introduced an interface to decouple the layers.

To reduce the entity loading as drastically as I did, it is necessary to use the cache where we can. This is one of the mechanisms. I feel it's an appropriate way to manage the dependencies between the design layers. The logic is part of the entity layer, as all shared logic is implemented as default methods on the interfaces.

Reply:
OK, I see. Let's see how it goes.